### PR TITLE
TTD-281: mm-gql filtering

### PR DESCRIPTION
**Pull request by allantractiontools**
- There are some failing tests but it's also in the mm-gql branch.
- ~~Need to check why the results are typed never this is also happening on the mm-gql branch. Added a comment ``@TODO_NEVER_TYPE_ISSUE``~~
- ~~Results are not updating when the filter is updated. I'm not sure if this is because of the mock data generator because when I checked on it yesterday using pinnacle backend the results were updating.~~

### CODE REVIEW

File: src/queriers.spec.ts [hunk 1]
```diff
@@ -146,7 +150,734 @@ test('query can query data using multiple tokens, by making parallel requests',
...
+
+  const { data } = await mmGQLInstance.query({
+    users: queryDefinition({
+      def: generateUserNode(mmGQLInstance),
+      map: ({ id, archived }) => ({
+        id,
+        archived,
+      }),
+      filter: {
+        archived: { _eq: false },
```
Comment (Ricardo-Marques) [inline src/queriers.spec.ts:505]: test message mentions `_neq` but you're using `_eq`

File: src/queriers.spec.ts [hunk 2]
```diff
@@ -146,7 +150,734 @@ test('query can query data using multiple tokens, by making parallel requests',
...
+          filter: {
+            task: { _contains: 'my todo' },
+          },
+        }),
+      }),
+    }),
+  });
+
+  // @ts-ignore-error
+  expect(data.users[0].todos.length).toBe(2);
```
Comment (Ricardo-Marques) [inline src/queriers.spec.ts:738]: Wouldn't expect this to need to be ignored^ is this potentially remnants of when you were getting that incorrectly inferred `never` type?

File: src/queriers.spec.ts [hunk 3]
```diff
@@ -146,7 +150,734 @@ test('query can query data using multiple tokens, by making parallel requests',
...
+      map: ({ id, score, todos }) => ({
+        id,
+        score,
+        todos: todos({
+          map: ({ id, users }) => ({
+            id,
+            users: users({
+              map: ({ firstName }) => ({ firstName }),
+              filter: {
+                firstName: { _contains: 'john' },
```
Comment (Ricardo-Marques) [inline src/queriers.spec.ts:785]: Nice, thank you for making this case insensitive.

File: src/queriers.spec.ts [hunk 4]
```diff
@@ -146,7 +150,734 @@ test('query can query data using multiple tokens, by making parallel requests',
...
+  });
+
+  try {
+    await mmGQLInstance.query({
+      users: queryDefinition({
+        def: generateUserNode(mmGQLInstance),
+        map: ({ id }) => ({
+          id,
+          // @ts-ignore-error
+          address: ({ state }) => ({ state }),
```
Comment (Ricardo-Marques) [inline src/queriers.spec.ts:863]: right syntax is `address: address({ map: ({ state }) => ({ state }) })` if you want to get rid of this ts error

File: src/queryDefinitionAdapters.ts [hunk 5]
```diff
@@ -450,7 +457,21 @@ function getIdsString(ids: Array<string>) {
 export function getKeyValueFilterString<TNode extends INode>(
   filter: ValidFilterForNode<TNode>
 ) {
-  const convertedToDotFormat = prepareObjectForBE(filter, {
+  /**
+   * @TODO MM-486: Currently SM only supports _eq condition when filtering.
+   * we need to filter out the other conditions for now.
+   */
+  const flattenedFilters = getFlattenedNodeFilterObject(filter);
```
Comment (Ricardo-Marques) [inline src/queryDefinitionAdapters.ts:464]: I don't know that the new API supports any type of filtering yet. 
Lets leave this fully client side for now (comments this out). 

I'll start discussing filtering and sorting needs with the BE team today and we can start adding actual BE sorting/filtering/pagination little by little to these query builders, as it becomes available.

When we do start adding server side filtering/pagination/sorting, I want to add a way to conditionally disable that and revert back to client side. Something like

```
const CLIENT_SIDE = 'CLIENT_SIDE'

const mmGQLClient = new MMGQL({
  filtering: CLIENT_SIDE,
  sorting: SERVER_SIDE
})
```

File: src/types.ts [hunk 6]
```diff
@@ -242,25 +242,39 @@ export type GetParsedValueTypeFromDefaultFn<
...
    * }
+   * 
+   * setting TFilter to true
+   * 
+   * will return 
+   * 
+   * {
+   *   flag: Record<FilterCondition, boolean>
+   *   name: Record<FilterCondition, string>
+   * }
```
Comment (Ricardo-Marques) [inline src/types.ts:253]: I don't love this, even though I understand why you would add it here, I'd rather have an additional type (even if it has some repetition) to build a valid filter from a record of IData properties

**Review by Ricardo-Marques — APPROVED** [review]
Fantastic work Allan. This is going to be a tremendous help in our development process

File: src/queriers.spec.ts (hunk 2 above)
Comment (allantractiontools) [reply src/queriers.spec.ts:738]: Yes, forgot to remove them.

File: src/queriers.spec.ts (hunk 1 above)
Comment (allantractiontools) [reply src/queriers.spec.ts:505]: Fixed. Thanks for catching this. This is my copy-paste problem 😆.

File: src/queryDefinitionAdapters.ts (hunk 5 above)
Comment (allantractiontools) [reply src/queryDefinitionAdapters.ts:464]: Yes, the current API supports it but only for equal conditions.
Query would look like. This is what we have in the mm-gql branch.
```
filter: {firstName: "Sample"}
```

I'll remove this for now and do all the filtering on the client-side.

Comment (allantractiontools) [reply src/queryDefinitionAdapters.ts:464]: Created a separate card
https://tractiontools.atlassian.net/browse/TTD-316

File: src/types.ts (hunk 6 above)
Comment (allantractiontools) [reply src/types.ts:253]: Created separate type for filters.

**Review by Ricardo-Marques — APPROVED** [review]

**Pull request merged**